File: lib/constants/bosses.ts
export interface BossDefinition {
  name: string;
  slug: string;
  raid: string;
  raidSlug: string;
  wowBossId?: number; // ENCOUNTER_START numeric ID (when present)
  aliases?: string[]; // alternative names seen in logs
  sortOrder: number;
}

// WotLK encounter IDs (Warmane / Blizzard WotLK)
// Ordered most-relevant first: ICC → Ruby Sanctum → ToC → Ulduar → VoA → EoE → OS → Naxx
export const WOTLK_BOSSES: BossDefinition[] = [
  // ─── ICECROWN CITADEL ────────────────────────────────────────
  { name: "Lord Marrowgar",       slug: "lord-marrowgar",       raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36612, sortOrder: 10 },
  { name: "Lady Deathwhisper",    slug: "lady-deathwhisper",    raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36855, sortOrder: 11 },
  { name: "Gunship Battle",       slug: "gunship-battle",       raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 37813, sortOrder: 12, aliases: ["Skybreaker", "Orgrim's Hammer"] },
  { name: "Deathbringer Saurfang",slug: "deathbringer-saurfang",raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 37813, sortOrder: 13 },
  { name: "Festergut",            slug: "festergut",            raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36626, sortOrder: 20 },
  { name: "Rotface",              slug: "rotface",              raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36627, sortOrder: 21 },
  { name: "Professor Putricide",  slug: "professor-putricide",  raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36678, sortOrder: 22 },
  { name: "Blood Prince Council", slug: "blood-prince-council", raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 37958, sortOrder: 30, aliases: ["Prince Valanar", "Prince Keleseth", "Prince Taldaram"] },
  { name: "Blood-Queen Lana'thel",slug: "blood-queen-lanathel", raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 37955, sortOrder: 31 },
  { name: "Valithria Dreamwalker",slug: "valithria-dreamwalker",raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36789, sortOrder: 32 },
  { name: "Sindragosa",           slug: "sindragosa",           raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36853, sortOrder: 33 },
  { name: "The Lich King",        slug: "the-lich-king",        raid: "Icecrown Citadel",           raidSlug: "icecrown-citadel",           wowBossId: 36597, sortOrder: 40, aliases: ["Lich King", "Arthas"] },

  // ─── RUBY SANCTUM ────────────────────────────────────────────
  { name: "Halion",               slug: "halion",               raid: "Ruby Sanctum",               raidSlug: "ruby-sanctum",               wowBossId: 39863, sortOrder: 100 },

  // ─── TRIAL OF THE CRUSADER ───────────────────────────────────
  { name: "Northrend Beasts",     slug: "northrend-beasts",     raid: "Trial of the Crusader",      raidSlug: "trial-of-the-crusader",      wowBossId: 34796, sortOrder: 200, aliases: ["Gormok the Impaler", "Icehowl", "Acidmaw", "Dreadscale"] },
  { name: "Lord Jaraxxus",        slug: "lord-jaraxxus",        raid: "Trial of the Crusader",      raidSlug: "trial-of-the-crusader",      wowBossId: 34780, sortOrder: 201 },
  { name: "Faction Champions",    slug: "faction-champions",    raid: "Trial of the Crusader",      raidSlug: "trial-of-the-crusader",      wowBossId: 34968, sortOrder: 202 },
  { name: "Twin Val'kyr",         slug: "twin-valkyr",          raid: "Trial of the Crusader",      raidSlug: "trial-of-the-crusader",      wowBossId: 34497, sortOrder: 203, aliases: ["Fjola Lightbane", "Eydis Darkbane"] },
  { name: "Anub'arak",            slug: "anubarak",             raid: "Trial of the Crusader",      raidSlug: "trial-of-the-crusader",      wowBossId: 34564, sortOrder: 204 },

  // ─── ULDUAR ──────────────────────────────────────────────────
  { name: "Flame Leviathan",      slug: "flame-leviathan",      raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33113, sortOrder: 300 },
  { name: "Ignis the Furnace Master", slug: "ignis",            raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33118, sortOrder: 301 },
  { name: "Razorscale",           slug: "razorscale",           raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33186, sortOrder: 302 },
  { name: "XT-002 Deconstructor", slug: "xt-002",               raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33293, sortOrder: 303 },
  { name: "Assembly of Iron",     slug: "assembly-of-iron",     raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33271, sortOrder: 310, aliases: ["Steelbreaker", "Molgeim", "Brundir"] },
  { name: "Kologarn",             slug: "kologarn",             raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 32930, sortOrder: 311 },
  { name: "Auriaya",              slug: "auriaya",              raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33515, sortOrder: 312 },
  { name: "Hodir",                slug: "hodir",                raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 32845, sortOrder: 320 },
  { name: "Thorim",               slug: "thorim",               raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 32865, sortOrder: 321 },
  { name: "Freya",                slug: "freya",                raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 32906, sortOrder: 322 },
  { name: "Mimiron",              slug: "mimiron",              raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33350, sortOrder: 323 },
  { name: "General Vezax",        slug: "general-vezax",        raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33271, sortOrder: 330 },
  { name: "Yogg-Saron",           slug: "yogg-saron",           raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 33288, sortOrder: 331 },
  { name: "Algalon the Observer", slug: "algalon",              raid: "Ulduar",                     raidSlug: "ulduar",                     wowBossId: 32871, sortOrder: 332 },

  // ─── VAULT OF ARCHAVON ───────────────────────────────────────
  { name: "Archavon the Stone Watcher", slug: "archavon",      raid: "Vault of Archavon",          raidSlug: "vault-of-archavon",          wowBossId: 31125, sortOrder: 400 },
  { name: "Emalon the Storm Watcher",   slug: "emalon",        raid: "Vault of Archavon",          raidSlug: "vault-of-archavon",          wowBossId: 33993, sortOrder: 401 },
  { name: "Koralon the Flame Watcher",  slug: "koralon",       raid: "Vault of Archavon",          raidSlug: "vault-of-archavon",          wowBossId: 35013, sortOrder: 402 },
  { name: "Toravon the Ice Watcher",    slug: "toravon",       raid: "Vault of Archavon",          raidSlug: "vault-of-archavon",          wowBossId: 38433, sortOrder: 403 },

  // ─── EYE OF ETERNITY ─────────────────────────────────────────
  { name: "Malygos",              slug: "malygos",              raid: "Eye of Eternity",            raidSlug: "eye-of-eternity",            wowBossId: 28859, sortOrder: 500 },

  // ─── OBSIDIAN SANCTUM ────────────────────────────────────────
  { name: "Sartharion",           slug: "sartharion",           raid: "The Obsidian Sanctum",       raidSlug: "obsidian-sanctum",           wowBossId: 28860, sortOrder: 600 },

  // ─── NAXXRAMAS ───────────────────────────────────────────────
  { name: "Anub'Rekhan",          slug: "anub-rekhan",          raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15956, sortOrder: 700 },
  { name: "Grand Widow Faerlina", slug: "grand-widow-faerlina", raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15953, sortOrder: 701 },
  { name: "Maexxna",              slug: "maexxna",              raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15952, sortOrder: 702 },
  { name: "Noth the Plaguebringer", slug: "noth-the-plaguebringer", raid: "Naxxramas",              raidSlug: "naxxramas",                  wowBossId: 15954, sortOrder: 710 },
  { name: "Heigan the Unclean",   slug: "heigan-the-unclean",   raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15936, sortOrder: 711 },
  { name: "Loatheb",              slug: "loatheb",              raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 16011, sortOrder: 712 },
  { name: "Instructor Razuvious", slug: "instructor-razuvious", raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 16061, sortOrder: 720 },
  { name: "Gothik the Harvester", slug: "gothik-the-harvester", raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 16060, sortOrder: 721 },
  { name: "The Four Horsemen",    slug: "the-four-horsemen",    raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 16062, sortOrder: 722, aliases: ["Baron Rivendare", "Highlord Mograine", "Lady Blaumeux", "Thane Kor'thazz"] },
  { name: "Patchwerk",            slug: "patchwerk",            raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 16028, sortOrder: 730 },
  { name: "Grobbulus",            slug: "grobbulus",            raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15931, sortOrder: 731 },
  { name: "Gluth",                slug: "gluth",                raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15932, sortOrder: 732 },
  { name: "Thaddius",             slug: "thaddius",             raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15928, sortOrder: 733 },
  { name: "Sapphiron",            slug: "sapphiron",            raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15989, sortOrder: 740 },
  { name: "Kel'Thuzad",           slug: "kelthuzad",            raid: "Naxxramas",                  raidSlug: "naxxramas",                  wowBossId: 15990, sortOrder: 741, aliases: ["Kel'Thuzad"] },
];

// Lookup maps for fast parser access
export const BOSS_BY_NAME = new Map(
  WOTLK_BOSSES.map(b => [b.name.toLowerCase(), b])
);

export const BOSS_BY_ID = new Map(
  WOTLK_BOSSES.filter(b => b.wowBossId).map(b => [b.wowBossId!, b])
);

export const BOSS_BY_SLUG = new Map(
  WOTLK_BOSSES.map(b => [b.slug, b])
);

// All known boss/alias names flattened for quick set lookup
export const ALL_BOSS_NAMES: Set<string> = new Set([
  ...WOTLK_BOSSES.map(b => b.name.toLowerCase()),
  ...WOTLK_BOSSES.flatMap(b => (b.aliases ?? []).map(a => a.toLowerCase())),
]);

export const RAIDS = [
  ...new Map(WOTLK_BOSSES.map(b => [b.raidSlug, { name: b.raid, slug: b.raidSlug }])).values(),
];

File: parser/bosses.py
"""
WotLK boss definitions — mirrors lib/constants/bosses.ts.
Used by the parser to detect raid encounters.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional

@dataclass
class BossDef:
    name: str
    slug: str
    raid: str
    wow_boss_id: Optional[int] = None
    aliases: list[str] = field(default_factory=list)


BOSSES: list[BossDef] = [
    # Naxxramas
    BossDef("Anub'Rekhan",           "anub-rekhan",          "Naxxramas",               15956),
    BossDef("Grand Widow Faerlina",  "grand-widow-faerlina", "Naxxramas",               15953),
    BossDef("Maexxna",               "maexxna",              "Naxxramas",               15952),
    BossDef("Noth the Plaguebringer","noth-the-plaguebringer","Naxxramas",              15954),
    BossDef("Heigan the Unclean",    "heigan-the-unclean",   "Naxxramas",               15936),
    BossDef("Loatheb",               "loatheb",              "Naxxramas",               16011),
    BossDef("Instructor Razuvious",  "instructor-razuvious", "Naxxramas",               16061),
    BossDef("Gothik the Harvester",  "gothik-the-harvester", "Naxxramas",               16060),
    BossDef("The Four Horsemen",     "the-four-horsemen",    "Naxxramas",               16062,
            ["Baron Rivendare", "Highlord Mograine", "Lady Blaumeux", "Thane Kor'thazz"]),
    BossDef("Patchwerk",             "patchwerk",            "Naxxramas",               16028),
    BossDef("Grobbulus",             "grobbulus",            "Naxxramas",               15931),
    BossDef("Gluth",                 "gluth",                "Naxxramas",               15932),
    BossDef("Thaddius",              "thaddius",             "Naxxramas",               15928),
    BossDef("Sapphiron",             "sapphiron",            "Naxxramas",               15989),
    BossDef("Kel'Thuzad",            "kelthuzad",            "Naxxramas",               15990),

    # Eye of Eternity
    BossDef("Malygos",               "malygos",              "Eye of Eternity",         28859),

    # Obsidian Sanctum
    BossDef("Sartharion",            "sartharion",           "The Obsidian Sanctum",    28860),

    # Vault of Archavon
    BossDef("Archavon the Stone Watcher","archavon",         "Vault of Archavon",       31125),
    BossDef("Emalon the Storm Watcher",  "emalon",           "Vault of Archavon",       33993),
    BossDef("Koralon the Flame Watcher", "koralon",          "Vault of Archavon",       35013),
    BossDef("Toravon the Ice Watcher",   "toravon",          "Vault of Archavon",       38433),

    # Ulduar
    BossDef("Flame Leviathan",       "flame-leviathan",      "Ulduar",                  33113),
    BossDef("Ignis the Furnace Master","ignis",              "Ulduar",                  33118),
    BossDef("Razorscale",            "razorscale",           "Ulduar",                  33186),
    BossDef("XT-002 Deconstructor",  "xt-002",               "Ulduar",                  33293),
    BossDef("Assembly of Iron",      "assembly-of-iron",     "Ulduar",                  None,
            ["Steelbreaker", "Molgeim", "Brundir"]),
    BossDef("Kologarn",              "kologarn",             "Ulduar",                  32930),
    BossDef("Auriaya",               "auriaya",              "Ulduar",                  33515),
    BossDef("Hodir",                 "hodir",                "Ulduar",                  32845),
    BossDef("Thorim",                "thorim",               "Ulduar",                  32865),
    BossDef("Freya",                 "freya",                "Ulduar",                  32906),
    BossDef("Mimiron",               "mimiron",              "Ulduar",                  33350),
    BossDef("General Vezax",         "general-vezax",        "Ulduar",                  33271),
    BossDef("Yogg-Saron",            "yogg-saron",           "Ulduar",                  33288),
    BossDef("Algalon the Observer",  "algalon",              "Ulduar",                  32871),

    # Trial of the Crusader
    BossDef("Northrend Beasts",      "northrend-beasts",     "Trial of the Crusader",   34796,
            ["Gormok the Impaler", "Icehowl", "Acidmaw", "Dreadscale"]),
    BossDef("Lord Jaraxxus",         "lord-jaraxxus",        "Trial of the Crusader",   34780),
    BossDef("Faction Champions",     "faction-champions",    "Trial of the Crusader",   34968),
    BossDef("Twin Val'kyr",          "twin-valkyr",          "Trial of the Crusader",   34497,
            ["Fjola Lightbane", "Eydis Darkbane"]),
    BossDef("Anub'arak",             "anubarak",             "Trial of the Crusader",   34564),

    # Icecrown Citadel
    BossDef("Lord Marrowgar",        "lord-marrowgar",       "Icecrown Citadel",        36612),
    BossDef("Lady Deathwhisper",     "lady-deathwhisper",    "Icecrown Citadel",        36855),
    BossDef("Gunship Battle",        "gunship-battle",       "Icecrown Citadel",        None,
            # Horde side: players jetpack to The Skybreaker and kill these crew members.
            # High Captain Justin Bartlett is the Alliance NPC whose death signals a kill.
            ["Muradin Bronzebeard", "High Captain Justin Bartlett",
             "Skybreaker Sorcerer", "Skybreaker Rifleman", "Skybreaker Sergeant"]),
    BossDef("Deathbringer Saurfang", "deathbringer-saurfang","Icecrown Citadel",        37813),
    BossDef("Festergut",             "festergut",            "Icecrown Citadel",        36626),
    BossDef("Rotface",               "rotface",              "Icecrown Citadel",        36627),
    BossDef("Professor Putricide",   "professor-putricide",  "Icecrown Citadel",        36678),
    BossDef("Blood Prince Council",  "blood-prince-council", "Icecrown Citadel",        37958,
            ["Prince Valanar", "Prince Keleseth", "Prince Taldaram"]),
    BossDef("Blood-Queen Lana'thel", "blood-queen-lanathel", "Icecrown Citadel",        37955),
    BossDef("Valithria Dreamwalker", "valithria-dreamwalker","Icecrown Citadel",        36789),
    BossDef("Sindragosa",            "sindragosa",           "Icecrown Citadel",        36853),
    BossDef("The Lich King",         "the-lich-king",        "Icecrown Citadel",        36597,
            ["Lich King", "Arthas"]),

    # Ruby Sanctum
    BossDef("Halion",                "halion",               "Ruby Sanctum",            39863),
]

# ── Lookup maps ──────────────────────────────────────────────────

# name (lower) → BossDef
BOSS_BY_NAME: dict[str, BossDef] = {}
for _b in BOSSES:
    BOSS_BY_NAME[_b.name.lower()] = _b
    for _alias in _b.aliases:
        BOSS_BY_NAME[_alias.lower()] = _b

# wow_boss_id → BossDef
BOSS_BY_ID: dict[int, BossDef] = {
    b.wow_boss_id: b for b in BOSSES if b.wow_boss_id
}

# flat set of all known boss/alias names
ALL_BOSS_NAMES: set[str] = set(BOSS_BY_NAME.keys())


def lookup_boss(name: str) -> Optional[BossDef]:
    return BOSS_BY_NAME.get(name.lower())


def lookup_boss_by_id(boss_id: int) -> Optional[BossDef]:
    return BOSS_BY_ID.get(boss_id)
